File: web/src/features/workspace/panes/agent/nativeChat.ts
export const nativeChatProviders = { codex: { name: 'Codex' } } as const
// Provider-neutral view model. Provider adapters live on the server; a future
// Claude adapter can emit this same contract without changing the chat surface.
export type ChatItem = {
  id: string
  kind: 'user' | 'assistant' | 'reasoning' | 'tool'
  text: string
  title?: string
  status?: string
}
export type ChatQuestion = {
  id: string
  header: string
  question: string
  isSecret?: boolean
  options?: { label: string; description: string }[]
}
export type ChatRequest = {
  id: number | string
  kind: 'approval' | 'input'
  title: string
  details: string
  questions?: ChatQuestion[]
}
export type ChatUsage = { inputTokens: number; outputTokens: number; totalTokens: number }
export type ChatState = {
  items: ChatItem[]
  requests: ChatRequest[]
  queuedMessages?: string[]
  working: boolean
  error?: string
  usage?: ChatUsage
  model?: string
  effort?: string
  sequence: number
}
export type ChatEvent =
  | { type: 'chat_snapshot' | 'chat_state'; sequence: number; value: Omit<ChatState, 'sequence'> }
  | { type: 'chat_item'; sequence: number; value: ChatItem }
  | { type: 'chat_error'; message: string }
  | { type: 'chat_sent' }
  | { type: 'chat_usage'; sequence: number; value: ChatUsage }
export const emptyChat: ChatState = { items: [], requests: [], working: false, sequence: -1 }
export function reduceChat(state: ChatState, event: ChatEvent): ChatState {
  if (event.type === 'chat_sent') return state
  if (event.type === 'chat_error') return { ...state, error: event.message }
  if (event.type !== 'chat_snapshot' && event.sequence <= state.sequence) return state
  if (event.type === 'chat_usage') return { ...state, usage: event.value, sequence: event.sequence }
  if (event.type === 'chat_item') {
    const index = state.items.findIndex((item) => item.id === event.value.id)
    const items = [...state.items]
    if (index < 0) items.push(event.value)
    else items[index] = event.value
    return { ...state, items, sequence: event.sequence }
  }
  return {
    ...event.value,
    items: event.value.items ?? [],
    requests: event.value.requests ?? [],
    sequence: event.sequence,
  }
}

File: internal/server/native_chat.go
package server

import "encoding/json"

// Native chat is a provider-neutral browser contract. Provider adapters own
// process startup, session persistence and conversion from their wire events.
type chatItem struct {
	ID     string `json:"id"`
	Kind   string `json:"kind"`
	Text   string `json:"text"`
	Title  string `json:"title,omitempty"`
	Status string `json:"status,omitempty"`
}
type chatRequest struct {
	ID        json.RawMessage `json:"id"`
	Kind      string          `json:"kind"`
	Title     string          `json:"title"`
	Details   string          `json:"details"`
	Questions json.RawMessage `json:"questions,omitempty"`
}
type chatUsage struct {
	InputTokens       int64 `json:"inputTokens"`
	OutputTokens      int64 `json:"outputTokens"`
	CachedInputTokens int64 `json:"cachedInputTokens"`
	TotalTokens       int64 `json:"totalTokens"`
}
type chatState struct {
	Items    []chatItem    `json:"items"`
	Requests []chatRequest `json:"requests"`
	Working  bool          `json:"working"`
	Error    string        `json:"error,omitempty"`
	Model    string        `json:"model,omitempty"`
	Effort   string        `json:"effort,omitempty"`
	Usage    *chatUsage    `json:"usage,omitempty"`
}
type chatClientMessage struct {
	Type      string                `json:"type"`
	Message   string                `json:"message"`
	Model     string                `json:"model"`
	Effort    string                `json:"effort"`
	Images    []piNativeClientImage `json:"images"`
	RequestID json.RawMessage       `json:"requestId"`
	Decision  string                `json:"decision"`
	Answers   map[string]struct {
		Answers []string `json:"answers"`
	} `json:"answers"`
}
